### ISSUE TITLE
shfm $PWD output doesn't act like stdout at all

### ISSUE THREAD

**Issue by dther** [dylanaraps/shfm#29]
shfm always prints to the screen, no matter how I call it. `shfm > file`, `cd "$(shfm)"`, `shfm() { cd "$(command shfm "$@")"; }` have all failed to do anything except print $PWD to the terminal. I've tried this using bash, dash, zsh and yash. Has anyone gotten this working at all?

EDIT: it inexplicably works on my server, but not on my daily driver. I have no idea why this is. The only reason I can currently fathom is that my server runs bash 5.0 and my current installation is bash 5.1.

**Comment by Sketch98** [comment]
i ran into the same problem with bash and zsh, but it works with dash. i wonder why that's the case.

**Comment by dther** [comment]
I tested it again and now it works just fine! I think this is what's happening: bash and zsh consider `/dev/tty` and stdout to be the same thing. In most cases, this is fine, since if you run a shell script, the contents of stdout are dumped to the screen. However, "technically", /dev/tty is supposed to be "What can be seen on the terminal" and stdout is, well, stdout. shfm relies on this little bit of pedantic, non-POSIX defined behaviour to write to the TUI to the terminal emulator while keeping it separate from the "actual" stdout (i.e. the directory on exit in plaintext). In dash, this works as intended. However, in bash/zsh, since `main` pipes straight into /dev/tty, *everything* gets piped to the terminal emulator, bypassing "normal" stdout entirely.

Between now and when I posted this issue, I've replaced my /bin/sh with dash. So if anyone has this problem: Change `#!/bin/sh` to `#!/bin/dash` on the first line, or link `/bin/sh` to `dash`. I might just replace `lf` yet. ;)

Closing this issue. Solutions are above. Alternatively, it might be a better idea to replace `/dev/tty` with `&2` (stderr) to avoid this issue entirely, and also avoids issues in other POSIX sh implementations. It seems to work in my tests, but I imagine Dylan had some internal reason for not using it...
